rust/rules.bzl

```diff
@@ -0,0 +1,117 @@
+load("@io_bazel_rules_rust//rust:rust.bzl", "rust_library")
+load("//protobuf:rules.bzl", "proto_compile", "proto_language_deps", "proto_repositories")
+load("//rust/raze:crates.bzl", "raze_fetch_remote_crates")
+
+def rust_proto_repositories(
+    lang_deps = {},
+    lang_requires = [],
+    **kwargs):
+  raze_fetch_remote_crates()
+
+def rust_proto_compile(langs = [str(Label("//rust"))], **kwargs):
+  proto_compile(langs = langs, **kwargs)
+
+PB_COMPILE_DEPS = [
+    "@org_pubref_rules_protobuf//rust/raze:protobuf",
+]
+
+GRPC_COMPILE_DEPS = PB_COMPILE_DEPS + [
+    "@org_pubref_rules_protobuf//rust/raze:grpc",
+    "@org_pubref_rules_protobuf//rust/raze:tls_api",
+    "@org_pubref_rules_protobuf//rust/raze:tls_api_stub",
+]
+
+def _basename(f):
+  return f.basename[:-len(f.extension) - 1]
+
+def _gen_lib_impl(ctx):
+  content = ["extern crate protobuf;"]
+  if ctx.attr.grpc:
+    content.append("extern crate grpc;")
+    content.append("extern crate tls_api;")
+  for dep in ctx.attr.deps:
+    content.append("extern crate %s;" % dep.label.name)
+    content.append("use %s::*;" % dep.label.name)
+  for f in ctx.files.files:
+    content.append("pub mod %s;" % _basename(f))
+    content.append("pub use %s::*;" % _basename(f))
+    if ctx.attr.grpc:
+      content.append("pub mod %s_grpc;" % _basename(f))
+      content.append("pub use %s_grpc::*;" % _basename(f))
+  ctx.actions.write(
+    ctx.outputs.out,
+    "\n".join(content),
```

Comment:
As an extension of the above content, I had to do some postprocessing to the generated protobuf content -- specifically replacing `super::` stuff. Is this not applicable for you?

Actually, as I look at this, I'm not seeing where the complied source even goes...

Reply:
The generated source goes into the same folder as the lib.rs. I don't need to replace supper:: stuff because I maintain the same lib structure. I don't know though how it will behave with some multi component package in proto.